File: packages/provide-uterm-server/src/provide/uterm/server/egress.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
"""Egress target validation for outbound connector connections."""

from __future__ import annotations

import asyncio
import ipaddress
import time
from typing import TYPE_CHECKING, Any
from urllib.parse import urlparse

from provide.uterm.server.webhooks import _METADATA_IPS, _resolve_host

if TYPE_CHECKING:
    from collections.abc import Mapping

# TTL for the per-host DNS cache used by assert_webhook_target_allowed.
_EGRESS_DNS_TTL_S: float = 60.0
# Bound the per-host DNS resolution so a slow / hostile resolver can't hang a
# session-create or webhook evaluation.  ``_resolve_host`` itself sets no
# timeout, so we wrap it in ``asyncio.wait_for`` at the guard call sites; the
# resulting ``TimeoutError`` is an ``OSError`` subclass and so flows through the
# fail-closed ``except OSError`` branch below.
_EGRESS_RESOLVE_TIMEOUT_S: float = 5.0
# host → (timestamp, resolved_addresses)
_resolve_cache: dict[str, tuple[float, tuple[str, ...]]] = {}

# NAT64 well-known prefix (RFC 6052).  Any address in this /96 carries an IPv4
# in its low 32 bits and, in a NAT64-enabled cluster, is actually translated to
# that IPv4 — so 64:ff9b::169.254.169.254 reaches the v4 metadata service.
_NAT64_WELL_KNOWN = ipaddress.ip_network("64:ff9b::/96")

# Residual risk — M3 (DNS rebinding), partially mitigated:
#   The connector guard validates the IP a hostname resolves to at *create*
#   time, but the connector itself re-resolves the hostname at *connect* time.
#   A name with a TTL-0 record that flips to a metadata / internal IP between
#   create and connect could otherwise reach it.  Pinning the literal IP into
#   connector_config and connecting to the IP is the WRONG fix: it breaks TLS
#   SNI (wss://) and SSH host-key/known-hosts verification, silently disabling
#   MITM protection.  Instead the SSH and WebSocket connectors validate the
#   *actual* connected peer IP (``assert_ip_allowed``) right after the transport
#   handshake completes and BEFORE any application/PTY data flows — the handshake
#   still uses the original hostname, so SNI / host-key verification is intact.
#   That closes the rebinding window (we validate the IP we actually reached).
#   The telnet connector cannot reach the peer IP without a public accessor on
#   its (cross-package) transport, so it remains create-time-only for now.
#   The webhook path is bounded by the 60s ``_resolve_cache`` TTL above (a
#   rebind is caught on the next cache miss).


def _decode_embedded_ipv4(ip: ipaddress.IPv6Address) -> ipaddress.IPv4Address | None:
    """Return the IPv4 address embedded in *ip*, for any embedding form, else None.

    Covers every IPv6 form that carries a reachable IPv4 so the metadata /
    private membership checks can be applied to the *decoded* address rather
    than the (membership-check-evading) IPv6 wrapper:

      * IPv4-mapped     ``::ffff:a.b.c.d``   -> ``ip.ipv4_mapped``
      * 6to4            ``2002:AABB:CCDD::``  -> ``ip.sixtofour``
      * NAT64 well-known ``64:ff9b::a.b.c.d`` -> low 32 bits (RFC 6052)
      * IPv4-compatible ``::a.b.c.d``         -> low 32 bits (deprecated)

    The deprecated IPv4-compatible form excludes ``::`` (unspecified) and
    ``::1`` (loopback), whose low bits are 0 / 1 and are handled by the normal
    IPv6 loopback / unspecified branches instead.
    """
    if ip.ipv4_mapped is not None:
        return ip.ipv4_mapped
    if ip.sixtofour is not None:
        return ip.sixtofour
    if ip in _NAT64_WELL_KNOWN:
        return ipaddress.IPv4Address(int(ip) & 0xFFFFFFFF)
    if (int(ip) >> 32) == 0 and (int(ip) & 0xFFFFFFFF) not in (0, 1):
        return ipaddress.IPv4Address(int(ip) & 0xFFFFFFFF)
    return None


async def _resolve_cached(host: str) -> tuple[str, ...]:
    """Return resolved addresses for *host*, using a TTL-bounded in-process cache.

    Caching prevents a DNS storm when the same governance webhook URL is
    consulted on every incoming message.  The 60-second TTL is short enough
    that a rebinding attack that switches a name to a metadata IP will be
    caught on the next cache miss.
    """
    now = time.time()
    cached = _resolve_cache.get(host)
    if cached is not None and (now - cached[0]) < _EGRESS_DNS_TTL_S:
        return cached[1]
    addrs = tuple(await asyncio.wait_for(_resolve_host(host), _EGRESS_RESOLVE_TIMEOUT_S))
    _resolve_cache[host] = (now, addrs)
    return addrs


async def assert_webhook_target_allowed(url: str) -> None:
    """Raise EgressBlockedError if the webhook URL host resolves to a cloud-metadata IP.

    Metadata IPs are never a legitimate webhook target.  Private/internal hosts
    ARE allowed — a policy engine may be internal; HTTPS + TLS cert validation
    (enforced at config-load) cover the rest.  DNS results are cached for
    ``_EGRESS_DNS_TTL_S`` seconds to avoid per-request resolver storms.
    """
    host = urlparse(url).hostname
    if not host:
        return
    h = host.strip().strip("[]")
    try:
        addresses: tuple[str, ...] = (str(ipaddress.ip_address(h)),)
    except ValueError:
        # A resolution OSError (incl. socket.gaierror / timeout) must fail closed
        # rather than propagate out as an HTTP 500 / hang the request.
        try:
            addresses = await _resolve_cached(h)
        except OSError:
            raise EgressBlockedError(f"webhook target {url!r} could not be resolved") from None
        # An empty resolve must fail closed (parity with the connector guard);
        # otherwise the loop below never runs and the URL is silently allowed.
        if not addresses:
            raise EgressBlockedError(f"webhook target {url!r} could not be resolved") from None
    for addr in addresses:
        ip = ipaddress.ip_address(addr)
        # Decode any embedded-IPv4 IPv6 form (mapped / 6to4 / NAT64 / compat) to
        # its IPv4 so a wrapped metadata address can't slip past the membership
        # check.  In a NAT64 cluster 64:ff9b::169.254.169.254 is reachable.
        if isinstance(ip, ipaddress.IPv6Address):
            embedded = _decode_embedded_ipv4(ip)
            if embedded is not None:
                ip = embedded
        if ip in _METADATA_IPS:
            raise EgressBlockedError(f"webhook target {url!r} resolves to a blocked metadata address")


class EgressBlockedError(ValueError):
    """Raised when a connector target resolves to a forbidden address."""


def _check_resolved_ip(
    ip: ipaddress.IPv4Address | ipaddress.IPv6Address,
    *,
    block_private: bool,
    on_metadata: str,
    on_private: str,
) -> None:
    """Apply the metadata-always / private-when-flag policy to one resolved IP.

    Decodes any embedded-IPv4 IPv6 form (mapped / 6to4 / NAT64 / compat) to its
    IPv4 so a wrapped metadata/private address can't slip past the membership
    checks.  Cloud-metadata IPs are ALWAYS blocked (raising ``EgressBlockedError``
    with the *on_metadata* message); private / loopback / link-local / reserved /
    multicast / unspecified are blocked only when *block_private* is True (raising
    with the *on_private* message).
    """
    if isinstance(ip, ipaddress.IPv6Address):
        embedded = _decode_embedded_ipv4(ip)
        if embedded is not None:
            ip = embedded
    if ip in _METADATA_IPS:
        raise EgressBlockedError(on_metadata)
    if block_private and (
        ip.is_private or ip.is_loopback or ip.is_link_local or ip.is_multicast or ip.is_unspecified or ip.is_reserved
    ):
        raise EgressBlockedError(on_private)


def assert_ip_allowed(ip_str: str, *, block_private: bool) -> None:
    """Validate an ALREADY-RESOLVED literal peer IP (no DNS resolution).

    Used by connectors for post-connect peer-IP validation (M3 DNS-rebinding
    mitigation): after the transport handshake completes — but before any
    application/PTY data flows — the connector reads the real peer IP and runs
    it through this check, aborting if it is a blocked target.  Because the IP
    is already a literal there is no second DNS lookup (and so no rebinding
    window), and the handshake still used the original hostname, so TLS SNI /
    cert validation and SSH host-key/known-hosts verification stay intact.

    Cloud-metadata IPs are ALWAYS blocked; private/internal ranges are blocked
    only when *block_private* is True.
    """
    ip = ipaddress.ip_address(ip_str.strip().strip("[]"))
    _check_resolved_ip(
        ip,
        block_private=block_private,
        on_metadata=f"connector peer {ip_str!r} is a blocked metadata address",
        on_private=f"connector peer {ip_str!r} is a blocked internal address",
    )


async def assert_connector_target_allowed(host: str, *, block_private: bool) -> None:
    """Validate a connector target host (literal IP or DNS name).

    Cloud-metadata IPs are ALWAYS blocked (never a legitimate terminal target).
    Private / loopback / link-local / reserved / multicast / unspecified are
    blocked only when *block_private* is True (multi-tenant / hosted posture);
    by default connectors may reach internal hosts, which is their purpose.
    Every address a DNS name resolves to is checked (defeats rebinding to a
    metadata IP at minimum).
    """
    h = host.strip().strip("[]")
    try:
        addresses: tuple[str, ...] = (str(ipaddress.ip_address(h)),)
    except ValueError:
        # A resolution OSError (incl. socket.gaierror / timeout) must fail closed
        # → EgressBlockedError (registry maps it to 422), not propagate as a 500.
        try:
            resolved = await asyncio.wait_for(_resolve_host(h), _EGRESS_RESOLVE_TIMEOUT_S)
        except OSError:
            raise EgressBlockedError(f"could not resolve connector host {host!r}") from None
        addresses = tuple(resolved)
        if not addresses:
            raise EgressBlockedError(f"could not resolve connector host {host!r}") from None
    for addr in addresses:
        # Decode any embedded-IPv4 IPv6 form (mapped / 6to4 / NAT64 / compat) and
        # apply the metadata (always) / private (when block_private) checks to
        # each resolved address — shared with the literal-IP ``assert_ip_allowed``.
        _check_resolved_ip(
            ipaddress.ip_address(addr),
            block_private=block_private,
            on_metadata=f"connector target {host!r} resolves to a blocked metadata address",
            on_private=f"connector target {host!r} resolves to a blocked internal address",
        )


async def assert_session_egress_allowed(
    connector_type: str,
    connector_config: Mapping[str, Any],
    *,
    block_private: bool,
) -> None:
    """Egress-guard a session's connector target, given its type and config.

    This is the single source of truth for deriving the outbound host from a
    connector type + config and applying ``assert_connector_target_allowed``.
    Called from the SessionRegistry chokepoint (so every create/update path is
    covered) and from the /api/connect route (defense-in-depth + synchronous
    422 mapping).

    Host derivation:
      * ssh / telnet  -> ``connector_config["host"]``
      * websocket     -> ``urlparse(connector_config["url"]).hostname`` (when a
                         ``url`` is present)

    Connector types that take no user-supplied host or URL (shell / local /
    pty / the internal tunnel websocket with no url) yield no host and are
    intentionally left unguarded — there is nothing attacker-controlled to
    validate.  When a host IS derived it is passed to
    ``assert_connector_target_allowed`` which raises ``EgressBlockedError``.
    """
    target_host: str | None = None
    if connector_type in {"ssh", "telnet"}:
        host = connector_config.get("host")
        target_host = str(host) if host is not None else None
    elif connector_type == "websocket":
        url = connector_config.get("url")
        if url is not None:
            target_host = urlparse(str(url)).hostname
    if target_host is not None:
        await assert_connector_target_allowed(target_host, block_private=block_private)

File: packages/provide-uterm-go/serverauth/egress.go
//
// SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
// SPDX-License-Identifier: AGPL-3.0-or-later
//

package serverauth

import (
	"context"
	"fmt"
	"net"
	"net/url"
	"strings"
)

// metadataIPs ports _net._METADATA_IPS — cloud-metadata service IPs that an
// outbound webhook must never reach.
var metadataIPs = []net.IP{
	net.ParseIP("169.254.169.254"),
	net.ParseIP("100.100.100.200"),
	net.ParseIP("fd00:ec2::254"),
}

// nat64WellKnown is the RFC 6052 NAT64 well-known prefix 64:ff9b::/96.
var _, nat64WellKnown, _ = net.ParseCIDR("64:ff9b::/96")

// EgressBlockedError ports egress.EgressBlockedError.
type EgressBlockedError struct{ msg string }

func (e *EgressBlockedError) Error() string { return e.msg }

// HostResolver resolves a hostname to IP strings; overridable in tests.
type HostResolver func(ctx context.Context, host string) ([]string, error)

func defaultResolver(ctx context.Context, host string) ([]string, error) {
	return net.DefaultResolver.LookupHost(ctx, host)
}

func isMetadataIP(ip net.IP) bool {
	// Decode any embedded-IPv4 IPv6 form so a wrapped metadata address can't
	// evade the membership check (mirrors _decode_embedded_ipv4).
	if decoded := decodeEmbeddedIPv4(ip); decoded != nil {
		ip = decoded
	}
	for _, m := range metadataIPs {
		if m != nil && m.Equal(ip) {
			return true
		}
	}
	return false
}

// decodeEmbeddedIPv4 ports egress._decode_embedded_ipv4 for the forms that
// carry a reachable IPv4: IPv4-mapped, 6to4, NAT64 well-known, and the
// deprecated IPv4-compatible form (excluding :: and ::1).
func decodeEmbeddedIPv4(ip net.IP) net.IP {
	if ip == nil {
		return nil
	}
	// IPv4-mapped ::ffff:a.b.c.d and plain IPv4 are already handled natively by
	// net.IP.Equal (which normalises 4-in-6), so To4() != nil needs no decode.
	if v4 := ip.To4(); v4 != nil {
		return nil
	}
	v6 := ip.To16()
	if v6 == nil {
		return nil
	}
	// 6to4 2002:AABB:CCDD::
	if v6[0] == 0x20 && v6[1] == 0x02 {
		return net.IPv4(v6[2], v6[3], v6[4], v6[5]).To4()
	}
	// NAT64 well-known 64:ff9b::/96
	if nat64WellKnown != nil && nat64WellKnown.Contains(ip) {
		return net.IPv4(v6[12], v6[13], v6[14], v6[15]).To4()
	}
	// IPv4-compatible ::a.b.c.d (high 96 bits zero), excluding :: and ::1.
	if isZeroPrefix(v6[:12]) {
		low := net.IPv4(v6[12], v6[13], v6[14], v6[15]).To4()
		excluded := v6[12] == 0 && v6[13] == 0 && v6[14] == 0 && (v6[15] == 0 || v6[15] == 1)
		if !excluded {
			return low
		}
	}
	return nil
}

func isZeroPrefix(b []byte) bool {
	for _, v := range b {
		if v != 0 {
			return false
		}
	}
	return true
}

// AssertWebhookTargetAllowed ports egress.assert_webhook_target_allowed: raise
// EgressBlockedError if the webhook URL host resolves to a cloud-metadata IP.
// Private/internal hosts ARE allowed. Resolution failure fails closed.
func AssertWebhookTargetAllowed(ctx context.Context, rawURL string, resolver HostResolver) error {
	if resolver == nil {
		resolver = defaultResolver
	}
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return nil
	}
	host := parsed.Hostname()
	if host == "" {
		return nil
	}
	h := strings.Trim(strings.TrimSpace(host), "[]")

	var addresses []string
	if literal := net.ParseIP(h); literal != nil {
		addresses = []string{literal.String()}
	} else {
		resolved, rerr := resolver(ctx, h)
		if rerr != nil {
			return &EgressBlockedError{fmt.Sprintf("webhook target %q could not be resolved", rawURL)}
		}
		if len(resolved) == 0 {
			return &EgressBlockedError{fmt.Sprintf("webhook target %q could not be resolved", rawURL)}
		}
		addresses = resolved
	}
	for _, addr := range addresses {
		ip := net.ParseIP(addr)
		if ip == nil {
			continue
		}
		if isMetadataIP(ip) {
			return &EgressBlockedError{
				fmt.Sprintf("webhook target %q resolves to a blocked metadata address", rawURL)}
		}
	}
	return nil
}
